src/sagemaker_training/modules.py

```diff
@@ -172,3 +184,29 @@ def import_module(uri, name=DEFAULT_MODULE_NAME):  # type: (str, str) -> module
         return module
     except Exception as e:  # pylint: disable=broad-except
         six.reraise(errors.ImportModuleError, errors.ImportModuleError(e), sys.exc_info()[2])
+
+
+def _get_codeartifact_index():
+    """
+    Build the authenticated codeartifact index url
+    https://docs.aws.amazon.com/codeartifact/latest/ug/python-configure-pip.html
+    :return: authenticated codeartifact index url
+    """
+    domain = os.getenv(CA_DOMAIN_ENV)
+    owner = os.getenv(CA_DOMAIN_OWNER_ENV)
+    repository = os.getenv(CA_REPOSITORY_ENV)
+    region = os.getenv(CA_REGION_ENV)
+    if domain is None or owner is None or repository is None or region is None:
+        raise Exception("not all code environment variables are set (CA_DOMAIN: {}, CA_DOMAIN_OWNER {}, CA_REPOSITORY: {}, CA_REGION: {})".format(domain, owner, repository, region))
+
+    logger.info("configuring pip to use codeartifact (domain: {}, domain owner {}, repository: {}, region: {})".format(domain, owner, repository, region))
+    try:
+        client = boto3.client("codeartifact", region_name=region)
+        auth_token_response = client.get_authorization_token(domain=domain, domainOwner=owner)
+        token = auth_token_response["authorizationToken"]
+        endpoint_response = client.get_repository_endpoint(domain=domain, domainOwner=owner, repository=repository, format="pypi")
+        unauthenticated_index = endpoint_response["repositoryEndpoint"]
+        return re.sub("https://", "https://aws:{}@".format(token), re.sub("{}/?$".format(repository), "{}/simple/".format(repository), unauthenticated_index))
```

Comment:
Please ensure https://aws works for all regions. Some snow-forted region may have different prefix. 

LGTM otherwise.

Reply:
+1. Have you tested this in China region? Are we sure it will not be like https://aws.cn?

Reply:
As per https://aws.amazon.com/codeartifact/faq/, CodeArtifact is not available everywhere

```
In which AWS Regions is CodeArtifact available?

CodeArtifact is available in the following 13 [AWS Regions](https://aws.amazon.com/about-aws/global-infrastructure/regional-product-services/):

US East (N.Virginia)
US East (Ohio)
US West (Oregon)
EU (Ireland)
EU (London)
EU (Frankfurt)
EU (Stockholm)
EU (Milan)
EU (Paris)
Asia Pacific (Sydney)
Asia Pacific (Tokyo)
Asia Pacific (Mumbai)
Asia Pacific (Singapore).
```

Additionally, the `aws:` part is not really abt the partition, it's the username e.g. `https://username:password@example.com/somewhere`

`client.get_repository_endpoint` returns the `https://example.com/somewhere`, then the following code is abt adding the username/password, and appending some path to the url